src/path.c

```diff
@@ -264,6 +264,66 @@ ly_path_check_predicate(const struct ly_ctx *ctx, const struct lysc_node *cur_no
     return LY_EVALID;
 }
 
+/**
+ * @brief Parse deref XPath function and perform all additional checks.
+ *
+ * @param[in] ctx libyang context.
+ * @param[in] ctx_node Optional context node, used for logging.
+ * @param[in] exp Parsed path.
+ * @param[in,out] tok_idx Index in @p exp, is adjusted.
+ * @return LY_ERR value.
+ */
+static LY_ERR
+ly_path_parse_deref(const struct ly_ctx *ctx, const struct lysc_node *ctx_node, const struct lyxp_expr *exp, uint32_t *tok_idx)
+{
+    LY_ERR ret = LY_SUCCESS;
+    size_t arg_len;
+    uint32_t begin_token, end_token;
+    struct lyxp_expr *arg_expr = NULL;
+
+    /* mandatory FunctionName */
+    LY_CHECK_ERR_GOTO(lyxp_next_token(ctx, exp, tok_idx, LYXP_TOKEN_FUNCNAME), ret = LY_EVALID, error);
+    if (exp->used <= *tok_idx + 1) {
+        if (ctx) {
+            LOGVAL(ctx, LY_VCODE_XP_EOF);
+        }
+        return LY_EINCOMPLETE;
+    }
+
+    if (!strncmp(&exp->expr[exp->tok_pos[*tok_idx]], "deref", 5)) {
+        LOGVAL(ctx, LYVE_XPATH, "Unexpected XPath function \"%.*s\" in path, expected \"deref(...)\"", exp->tok_len[*tok_idx], exp->tok_pos[*tok_idx]);
+        ret = LY_EVALID;
+        goto error;
+    }
+
+    /* mandatory '(' */
+    LY_CHECK_ERR_GOTO(lyxp_next_token(ctx, exp, tok_idx, LYXP_TOKEN_PAR1), ret = LY_EVALID, error);
+    begin_token = *tok_idx;
+
+    /* count tokens till ')' */
+    while (lyxp_check_token(NULL, exp, *tok_idx, LYXP_TOKEN_PAR2) && lyxp_check_token(NULL, exp, *tok_idx, LYXP_TOKEN_FUNCNAME)) {
+        (*tok_idx)++;
+    }
```

Comment:
What we talked about in some other conversation was parsing the argument as a nested path because that is what is allowed there so I would expect to see `lyd_parse_path()` here. The compile function expects a valid path but this does not check anything, pretty much.

Reply:
This is being already done, see lines 319-322. In your highlighted part I am just checking where the deref-arg path ends and that we don't have some nested XPath function within, therefore the closing ')' must be the one related to our deref-arg.

Reply:
Okay, will check properly after you address the other comment. Like I said, I have not executed the code yet so I may have read the code wrong.